lib/SILGen/SILGenApply.cpp

```diff
@@ -1118,7 +1118,11 @@ class SILGenApply : public Lowering::ExprVisitor<SILGenApply> {
     // Otherwise, we have a statically-dispatched call.
     auto constant = SILDeclRef(e->getDecl());
     if (e->getDecl()->getAttrs().hasAttribute<DistributedActorAttr>()) {
-      constant = constant.asDistributed(true);
+      // we're calling a distributed function, only in cross-actor situations
+      // does this actually need to call the thunk.
+      if (!selfApply) {
+        constant = constant.asDistributed(true);
```

Comment:
This is re-computing information that the type checker is already has, and with an algorithm that is likely to end up being different. Rather than check the `DistributedActorAttr` and `selfApply`, we should record in the call expression itself whether it's a potentially-remote call and therefore needs to go through the distributed thunk.

Reply:
I was concerned about that actually! Okey I'll ask you a bit in detail how to do that...

Reply:
Ok this is actually wrong, I unlocked a bunch of tests and realized that..

Reply:
This was wrong entirely, fix here: https://github.com/apple/swift/pull/39542